### DISCUSSION TITLE
Duplicate search results

### DISCUSSION THREAD

**Discussion by igbjnI05bF** [music-assistant/support#5353] [Feature requests and ideas]
### What version of Music Assistant has the issue?

2.2.2

### What version of the Home Assistant Integration have you got installed?

2024.8.1

### Have you tried everything in the Troubleshooting FAQ and reviewed the Open and Closed Issues and Discussions to resolve this yourself?

- [X] Yes

### The problem

Every search result shows duplicate entries. I have the Jellyfin plugin installed. When I search an artist, I see the artist twice - once with the musicassistant icon and one with the Jellyfin icon. When I hover over the musicassistant icon it says "this item is available in the library". When I hover over the jellyfin icon it says "jellyfin media server library".  Every artist, album, and song does this. 

### How to reproduce

Search for anything

### Music Providers

Music Assistant, RadioBrowser, Tune-In, and Jellyfin. I'm not even really sure what the MusicAssistant provider does, I can't disable it though, it's grayed out.

### Player Providers

Snapcast

### What version of Home Assistant Core are your running

2024.8.3

### What type of installation are you running?

Home Assistant OS

### On what type of hardware are you running?

Linux

**Comment by OzGav** [comment]
Can you provide a screenshot as the artist view search does not show provider icons. Perhaps you are referring to the global search?

**Comment by OzGav** [comment]
Looking at the global search I think that is what you are referring to. The GLOBAL search does what it says, it searches across all your providers for a result. So it makes sense to show that the artist is available in your library and in other providers. Why is this an issue?

**Comment by igbjnI05bF** [comment]
When I open Home Assistant I click on Music Assistant and then there is a big button for search at the bottom. Not sure if that's GLOBAL search or not, it doesn't say global.

The issue I have is that my music library, in my case Jellyfin, doesn't have two of everything, it has one of everything. When I, or anybody in my family, searches for anything and see duplicates it is cluttered at best and confusing at worst.
![IMG_20240829_074939](https://github.com/user-attachments/assets/5b57b353-8145-4650-8002-4988003b94fd)

**Comment by OzGav** [comment]
Ok then don’t use the global search. Go to Artist and search for Radiohead there.

**Comment by OzGav** [comment]
Also because you are using a mobile phone in portrait some info has to be removed. If you are on a wider screen device you will see all the provider icons,

**Comment by OzGav** [comment]
Additionally those results will do different things depending on settings for the queue controller when you select PLAY NOW. For the in library artist result you could have it just play the in library tracks but the Jellyfin result would play all tracks by the artist on your Jellyfin server.  So there are many reasons why having results displayed from all providers is preferable.

**Comment by igbjnI05bF** [comment]
Here is what is returned specifically for Artists and albums. As noted in original post, the "libraries" it says these are in are Jellyfin and then just "this item is available in the library". I don't know what this means.

As a test I disabled Jellyfin as a source. Now the artist Radiohead only shows up once. But now it is useless as when I click on it, it no longer takes me to Radiohead, it gives me three options "remove from library", "add to favorites", or "refresh item".

So clearly I don't have a separate library with Radiohead in it, if I did it would take me to the albums and let me start playing them. The only library that I have in MusicAssistant that actually has Radiohead is Jellyfin. So as an end-user I would expect, based on years of using searches online in various capacities, to not see duplicates of everything.

I appreciate the attention on this, but unless I'm misunderstanding something, it sounds like there is an programming reason why having duplicates makes sense to you, and this is a programmer vs end-user experience debate. I can assure you that no end user wants to see duplicates of *everything* every time they search something

If I'm missing something let me know sometimes I can be dumb
![20240829_08h18m23s_grim](https://github.com/user-attachments/assets/fed2edf9-71e9-48a0-bc7e-915465c82a4d)

**Comment by OzGav** [comment]
Sounds like you haven’t read the docs to understand how MA works https://music-assistant.io/usage/

You aren't seeing duplicates as a duplicate by definition is two of the same thing. These aren’t the same thing. One is showing you what you have added to the MA library (which in your case might well be everything in your Jellyfin library but this is not always the case) and what is available on the Jellyfin PROVIDER. The library and the source providers are different.

**Comment by OzGav** [comment]
Since it looks like you only have one provider then don’t use the global search. Just search on the individual views. You can hide the search icon in the settings.

**Comment by igbjnI05bF** [comment]
I will absolutely read through that link you posted. But before I do, I think it's important for software devs to understand the typical end user, which is *not* somebody is going to read up in detail every piece of software they use. After I read through that link, I may understand what you're saying and understand why it happens, but that doesn't mean that it's the best way to do it. My wife and children who use this system are never going to read through that. They are just going to search for artists or songs and they're going to constantly see what they interpret as duplicates, even if that isn't what you would call them.

I did find another search option by going to browse, selecting Jellyfin Media Server Library, Artists, and searching Radiohead there. Nothing is found. If I search for Arcade Fire, it does find it. It appears that there is some sort of lazy loading going on and it won't find the artist unless it has loaded in. So that search is useless. And even if it did work, it's many extra clicks to get to rather than the place you call "global search" but is just labeled as "search". 
![image](https://github.com/user-attachments/assets/dea07f3c-9070-443b-a2a8-f35e0f2eafd4)

**Comment by igbjnI05bF** [comment]
I have read through that link now. From what I can tell, it is saying that it adds everything in your sources to a MusicAssistant library, and then when you "search" (again, what you are calling "global search") it shows the MusicAssistant library as well as the Jellyfin source library. Ok, so I understand what's going on now. I think that is a ridiculous thing to do.

When I have a Nextcloud server and point it to a NFS share for files and it adds those NFS files to its database "library", when I search for a file does it show me both the original NFS share as well as the fact that it is now in the Nextcloud database? No, because why would I want to see the file twice lol. It searches the database which will then point to the NFS share, and the user doesn't need to be aware of any of it, they just search for and see the file.

If I'm somehow in the minority on this, and everybody else in the community loves the way this flow works, then I would ask for just a feature request to control what that "global search" button shows as results. But I can't imagine that I'm the only one that feels this way, because the way it currently is is very unintuitive for the layman.

**Comment by igbjnI05bF** [comment]
Also I would like to add that despite my strong feelings on this, Music Assistant is awesome and I appreciate the software and your time reading my rants

**Comment by OzGav** [comment]
Your Nextcloud example is not an apples and apples comparison but is does raise a  question. Why are you even using the Jellyfin provider? Why not just use the local provider? If you did then that would be the same as your Nextcloud example and would work as you expect.

**Comment by OzGav** [comment]
> From what I can tell, it is saying that it adds everything in your sources to a MusicAssistant library

No that is not what happens. Items added to your MA library depend upon the provider. For example, streaming providers only add what has been favourited (or similar term). Local providers (which includes Jellyfin) currently import all items into the MA library.

As an aside don’t use the browse view. That view causes more issues than it resolves and we regularly discuss removing it. It has its use case but not in your situation. As I mentioned above you should just being going to the Artist view and searching from there. This is because you only have one local provider. In the album and track views you should make sure you turn on the library only filters in those views.

**Comment by smcmnz** [comment]
I am having the same issue on 2.3.0b8 with a single file system provider loading music mounted to a docker container.

Duplicate results in global search appear in the Artists, Tracks and Albums panes but not the top results pane. Sometimes the album results have 3x results where the artist name and album name are the same. The results all link to the same underlying ID for the resource, and I have checked the DB and there is a single entry for each artist and album with that ID.

Running the same search in the artist and album view, the artist returns a single result, but albums return 2x results where artist = album.

I was also getting this on 2.2.2 and did a full DB reset when upgrading to remove duplicates. Best I can tell the DB contains only 1 result and the FE points duplicate results to the same ID.

**Comment by OzGav** [comment]
@smcmnz You do not have the same issue as the provider is different. Open a Q&A discussion with screenshots of what you are talking about

**Comment by OzGav** [comment]
@igbjnI05bF have you tried just using the local provider? Note that your files need to be tagged properly as per the docs

**Comment by igbjnI05bF** [comment]
> Your Nextcloud example is not an apples and apples comparison but is does raise a question. Why are you even using the Jellyfin provider? Why not just use the local provider? If you did then that would be the same as your Nextcloud example and would work as you expect.

I didn't use the Filesystem (remote share) because when I click on it it says "Support for music files that are present on remote SMB/CIFS" but I do not currently have a SMB/CIFS share, everything in my house is Linux so I have NFS. I'm surprised there isn't an NFS option.

I thought about just mounting the NFS share to a directory on the VM and just using the "Filesystem (local disk)" option but I'm running HAOS  which to my knowledge doesn't allow a traditional SSH in, or at least doesn't want you to. I do know there is a terminal app and maybe I can finagle it... but ultimately just decided to play ball and just set up the Jellyfin provider since it was the least amount of work rather than learning/implementing CIFS, or doing something HAOS doesn't really want you to do.

**Comment by igbjnI05bF** [comment]
> Additionally those results will do different things depending on settings for the queue controller when you select PLAY NOW. For the in library artist result you could have it just play the in library tracks but the Jellyfin result would play all tracks by the artist on your Jellyfin server. So there are many reasons why having results displayed from all providers is preferable.

They have different options? Not from what I'm seeing. I'm searching radiohead again, when I click on the three vertical dots, both the Radiohead for the MusicAssistant icon and the Jellyfin icon results show the same options: "Show info", "Add to library", or "Add to favorites". When I just click on the artist itself rather than the three dots, it just takes me to the artist.

Man I'm begging you to take your engineer hat off here. You're trying to justify to me why when you search for an artist or an album that you see results twice. For everything. I'm sure you're right... ther'es some engineer reason why this decision was made, but take that engineer hat off. Put yourself into the "wife approval" factor. This looks cluttered and like it is beta. This looks like a bug, even if you're saying it isn't a bug, 99% of non-engineers that just want to play music in their house are going to think this is a bug.

You're telling me to stop searching the way I'm searching and to search a different way, by clicking on artist first and then finding the much smaller search icon at the top right rather than the big search icon at the bottom. This is like Apple telling the world that the reason their calls are dropping is because they're holding the phone wrong. And guess what, I *still* can't find the search you're mentioning on mobile, which is the only place my family will ever use- when I ignore the big glorious "Search" button at the bottom and click on library -> artists, I don't see a different search button there. When my wife is in the kitchen and wants to listen to Gangster's Paradise while cooking a meatloaf she isn't going to walk upstairs to her office, get on her desktop, open a browser, go to homeassistant.ThatCrabbyComplainerGuyOnGithub.com, sign in, click "Music Assistant", remember to ignore the big glorius search button, click artists, then click on the much smaller search button on the top right, select kitchen speakers, Gangster's Paradise, then walk back downstairs to the kitchen.

No, she's going to open up her phone, open the app, click the big glorious "Search" button, search Gangster's Paradise, see duplicate results and think "why does my husband insist on using weird stuff instead of just sonos & spotify"

Listen man, you're a contributor to great software. I guarantee you're smarter than me. You're doing more for open source than I am. But engineers get stuck in an engineer mindset and don't think of the common person. If you would think of the common person, we wouldn't need 20 responses to a post of "hey maybe we shouldn't show duplicates for the jellyfin provider" lol. 


![20240830_13h50m44s_grim](https://github.com/user-attachments/assets/3c9068ad-ba74-4780-8220-e0c2411d114b)

**Comment by marcelveldt** [comment]
While I do agree that we could do a better job by at least trying to filter the duplicate results, I am not agreeing with all the sentiment in this ticket! I have never ever heard my wife or children be upset that they spotted some duplicate results in the global search, not even once. 

So with my engineer hat on: We have a very large project with so many way more important things to do which is why this did not get any priority YET, it will at some point. Keep in mind that we just are a handful weirdos spending our free time creating this software. 

You made your point and yes this will get looked at at some point but please dim down a bit please. I think you are filling in some blanks already for your family members ? Did you ask them if its really the end of the world to see some duplicate search results for a product that you can use for free ?

have a great weekend!

**Comment by igbjnI05bF** [comment]
I appreciate the response.

I do get pushback from friends/family when I try to introduce open source software rather than the mainstream proprietary alternatives. They do notice and point out and complain about the unpolished areas that stick out compared to the alternatives they would use as their first choice. In my experience there very much is an engineer-first mindset in FOSS which contributes to the situation, and saying "you're searching wrong" when I point out that the Jellyfin source has duplicate search results for everything did send me into rant mode. Your response is the first to acknowledge the issue, as minor as it might be.

Thanks again, I do love the software.

**Comment by OzGav** [comment]
> I'm surprised there isn't an NFS option.

If you are using the addon then people have made the NFS Mount in HA at Settings>>System>>Storage and then add network storage. You can then add it to MA as a Local Disk

**Comment by apophisnow** [comment]
@igbjnI05bF I just wanted to let you know that MA supports NFS now (at least in the beta), I don't know that it helps with the duplicates issue you brought up, but perhaps it will?
